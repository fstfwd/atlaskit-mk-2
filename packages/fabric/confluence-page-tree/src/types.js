// @flow

export type PageData = {
  id: string,
  title: string,
  childTypes: {
    page: {
      value: string,
    },
  },
};
export type PageAttribute = {
  id: string,
  history: any,
};

export type pageAttributesArgs = {
  contentIds: Array<string>,
  attributes: string,
  limit: number,
};

export type recursiveFetchArgs = {
  currentFetch: Promise<any>,
  terminatingFn: (response: any) => boolean,
  getNextFetch: () => Promise<any>,
  accumulator?: Array<any>,
};
