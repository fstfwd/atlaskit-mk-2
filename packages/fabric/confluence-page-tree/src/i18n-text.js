const i18n = {
  'en-au': {
    error: {
      title: 'Error text',
      description: 'Error description',
    },
    noaccess: {
      title: 'Well, this is awkward...',
      description:
        "You don't have access to these pages. Please speak to your site admin for help.",
    },
    empty: {
      title: "There's nothing here but potential",
      description:
        'Create some meeting notes, product requirements, decisions, or other content to see this space fill up.',
    },
    create: 'Create Page',
    tableHeaderTitle: 'Title',
    tableHeaderContributors: 'Contributors',
    tableHeaderLastModified: 'Last Modified',
    viewProfile: 'View Profile',
  },
};

function getI18n(lang = 'en-au') {
  return i18n[lang];
}

export { getI18n };
