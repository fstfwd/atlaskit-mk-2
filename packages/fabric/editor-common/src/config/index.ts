export * from './context-identifier-provider';
