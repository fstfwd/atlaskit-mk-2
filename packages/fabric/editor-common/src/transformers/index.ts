export * from './jira';
export { default as JIRATransformer } from './jira';

export * from './json';
export { default as JSONTransformer } from './json';

export * from './bitbucket';
export { default as BitbucketTransformer } from './bitbucket';

export * from './confluence';
export {
  default as ConfluenceTransformer,
  LANGUAGE_MAP as CONFlUENCE_LANGUAGE_MAP
} from './confluence';

export { Transformer } from './transformer';
