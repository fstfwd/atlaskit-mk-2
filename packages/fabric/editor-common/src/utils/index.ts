export * from './url';
export * from './uuid';
export * from './confluence';
export * from './languageList';
export { default as browser } from './browser';
export {
    default as ErrorReporter,
    ErrorReportingHandler,
} from './error-reporter';

export const rgbToHex = (value: string): string | undefined => {
  const matches = value.match(/(0?\.?\d{1,3})%?\b/g);
  if (matches && matches.length >= 3) {
    const [red, green, blue] = matches.map(Number);
    // tslint:disable-next-line:no-bitwise
    return '#' + ((blue | green << 8 | red << 16) | 1 << 24).toString(16).slice(1);
  }
};

/**
 * A replacement for `String.repeat` until it becomes widely available.
 */
export const stringRepeat = (text: string, length: number): string => {
  let result = '';
  for (let x = 0; x < length; x++) {
    result += text;
  }
  return result;
};
