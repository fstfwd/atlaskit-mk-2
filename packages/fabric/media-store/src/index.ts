export {
  MediaStore,
  MediaStoreConfig,
  MediaStoreGetFileImageParams,
} from './media-store';
export { MediaFile } from './models/media';
export { uploadFile } from './uploader';
