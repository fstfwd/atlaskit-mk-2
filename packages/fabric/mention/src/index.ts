import MentionResource, {
  AbstractMentionResource,
  MentionProvider,
} from './api/MentionResource';
import PresenceResource, {
  PresenceProvider,
  AbstractPresenceResource,
} from './api/PresenceResource';
import MentionItem from './components/MentionItem';
import MentionList from './components/MentionList';
import ResourcedMentionList from './components/ResourcedMentionList';
import MentionPicker from './components/MentionPicker';
import Mention from './components/Mention';
import ResourcedMention from './components/Mention/ResourcedMention';
import { MentionDescription, MentionsResult, isSpecialMention } from './types';
import { SearchIndex } from './util/searchIndex';
import { testData, storyData } from './support';
import * as support from './support';

export {
  // Classes
  MentionResource,
  PresenceResource,
  AbstractMentionResource,
  AbstractPresenceResource,
  SearchIndex,
  // Interfaces
  MentionProvider,
  PresenceProvider,
  MentionDescription,
  MentionsResult,
  // Components
  MentionItem,
  MentionList,
  ResourcedMentionList,
  MentionPicker,
  Mention,
  ResourcedMention,
  // Functions
  isSpecialMention,
  // Support data
  support,
  testData,
  storyData,
};

export default MentionPicker;
