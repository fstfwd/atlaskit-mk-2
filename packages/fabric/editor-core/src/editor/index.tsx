import * as React from 'react';
import { PropTypes } from 'react';
import { withAnalytics } from '@atlaskit/analytics';
import { createEditor, getUiComponent } from './create-editor';
import { createPluginsList } from './create-editor';
import EditorActions from './actions';
import ProviderFactory from '../providerFactory';
import { EditorProps, EditorInstance } from './types';
import { moveCursorToTheEnd } from '../utils';
export * from './types';

export interface State {
  editor?: EditorInstance;
}

export default class Editor extends React.Component<EditorProps, State> {
  static defaultProps: EditorProps = {
    appearance: 'message',
    disabled: false
  };

  static contextTypes = {
    editorActions: PropTypes.object
  };

  context: {
    editorActions?: EditorActions;
  };

  private providerFactory: ProviderFactory;

  constructor(props: EditorProps) {
    super(props);
    this.providerFactory = new ProviderFactory();
    this.state = {};
  }

  componentDidMount() {
    this.handleProviders(this.props);
  }

  componentWillReceiveProps(nextProps: EditorProps) {
    this.handleProviders(nextProps);
  }

  componentWillUnmount() {
    if (!this.state.editor) {
      return;
    }

    const { editor } = this.state;
    const { editorView } = editor;
    const { state: editorState } = editorView;

    this.unregisterEditorFromActions();

    editorState.plugins.forEach(plugin => {
      const state = plugin.getState(editor.editorView.state);
      if (state && state.destroy) {
        state.destroy();
      };
    });

    editorView.destroy();

    if (editor.eventDispatcher) {
      editor.eventDispatcher.destroy();
    }
  }

  componentDidUpdate(prevProps, prevState) {
    const { editor } = this.state;
    // Once the editor has been set for the first time
    if (!prevState.editor && editor) {
      // Focus editor first time we create it if shouldFocus prop is set to true.
      if (this.props.shouldFocus) {
        if (!editor.editorView.hasFocus()) {
          editor.editorView.focus();
        }

        moveCursorToTheEnd(editor.editorView);
      }
    }
    if (editor) {
      if (!editor.editorView.hasFocus()) {
        editor.editorView.focus()
      }
      moveCursorToTheEnd(editor.editorView);
    }
  }

  private registerEditorForActions(editor: EditorInstance) {
    if (this.context && this.context.editorActions) {
      this.context.editorActions._privateRegisterEditor(editor.editorView, editor.contentTransformer);
    }
  }

  private unregisterEditorFromActions() {
    if (this.context && this.context.editorActions) {
      this.context.editorActions._privateUnregisterEditor();
    }
  }

  private initEditor = place => {
    if (!place) {
      return;
    }
    const plugins = createPluginsList(this.props);
    const editor = createEditor(place, plugins, this.props, this.providerFactory);
    this.registerEditorForActions(editor);
    this.setState({ editor });
  }

  private handleProviders(props: EditorProps) {
    const {
      emojiProvider,
      mentionProvider,
      mediaProvider,
      collabEditProvider,
      activityProvider,
      presenceProvider,
      macroProvider,
      legacyImageUploadProvider
    } = props;
    this.providerFactory.setProvider('emojiProvider', emojiProvider);
    this.providerFactory.setProvider('mentionProvider', mentionProvider);
    this.providerFactory.setProvider('mediaProvider', mediaProvider);
    this.providerFactory.setProvider('imageUploadProvider', legacyImageUploadProvider);
    this.providerFactory.setProvider('collabEditProvider', collabEditProvider);
    this.providerFactory.setProvider('activityProvider', activityProvider);
    this.providerFactory.setProvider('presenceProvider', presenceProvider);
    this.providerFactory.setProvider('macroProvider', macroProvider);
  }

  render() {
    // tslint:disable-next-line:variable-name
    const { editor = {} } = this.state;

    const Component = getUiComponent(this.props.appearance);

    if (!Component) {
      return null;
    }

    const {
      editorView,
      contentComponents,
      primaryToolbarComponents,
      secondaryToolbarComponents,
      eventDispatcher
    } = editor as EditorInstance;

    return (
      <Component
        onUiReady={this.initEditor}

        disabled={this.props.disabled}

        editorView={editorView}
        providerFactory={this.providerFactory}

        eventDispatcher={eventDispatcher}

        maxHeight={this.props.maxHeight}
        onSave={this.props.onSave}
        onCancel={this.props.onCancel}

        popupsMountPoint={this.props.popupsMountPoint}
        popupsBoundariesElement={this.props.popupsBoundariesElement}

        contentComponents={contentComponents}
        primaryToolbarComponents={primaryToolbarComponents}
        secondaryToolbarComponents={secondaryToolbarComponents}

        customContentComponents={this.props.contentComponents}
        customPrimaryToolbarComponents={this.props.primaryToolbarComponents}
        customSecondaryToolbarComponents={this.props.secondaryToolbarComponents}

        addonToolbarComponents={this.props.addonToolbarComponents}
      />
    );
  }
}

export const EditorWithAnalytics = withAnalytics<typeof Editor>(Editor, {}, {}, true);
