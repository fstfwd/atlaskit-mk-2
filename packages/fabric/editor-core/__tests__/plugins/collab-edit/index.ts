import { EventEmitter } from 'events';
import { DecorationSet } from 'prosemirror-view';
import { Node as PmNode } from 'prosemirror-model';
import {
  doc,
  p,
  jestProsemirrorSerializer,
  jsdomFixtures,
} from '@atlaskit/editor-test-helpers';

import { setTextSelection } from '../../../src';
import createEditor from '../../_helpers/create-editor';
import ProviderFactory from '../../../src/providerFactory';
import collabEdit, {
  pluginKey as collabEditPluginKey,
} from '../../../src/editor/plugins/collab-edit';

// Telepointers information to the snapshot { doc: {...}, selection: {...}, telepointers: [...]}
expect.addSnapshotSerializer(
  jestProsemirrorSerializer(state => {
    const { decorations } = collabEditPluginKey.getState(state);
    const telepointers = (decorations as DecorationSet)
      .find(0, state.doc.nodeSize)
      // Latest version of PM exposes `from` & `to`
      .map(({ from, to, spec: { pointer: { sessionId } } }: any) => ({
        from,
        to,
        sessionId,
      }));
    return { telepointers };
  }),
);

class MockProvider extends EventEmitter {
  initialize() {}
  send() {}
}

const setupCollabEditor = (defaultValue?: PmNode | string | Object) => {
  const providerFactory = new ProviderFactory();
  const mockProvider: any = new MockProvider();
  const providerPromise = new Promise<MockProvider>(resolve =>
    resolve(mockProvider),
  );

  providerFactory.setProvider('collabEditProvider', providerPromise);
  const { editorView } = createEditor(
    [collabEdit],
    { defaultValue },
    providerFactory,
  );

  // Work around JSDOM/Node not supporting DOM Selection API
  if (
    !('getSelection' in window) &&
    navigator.userAgent.indexOf('Node.js') !== -1
  ) {
    jsdomFixtures(editorView);
  }

  const { refs } = editorView.state.doc as any;
  if (refs) {
    // Collapsed selection.
    if ('<>' in refs) {
      setTextSelection(editorView, refs['<>']);
      // Expanded selection
    } else if ('<' in refs || '>' in refs) {
      if ('<' in refs === false) {
        throw new Error('A `<` ref must complement a `>` ref.');
      }
      if ('>' in refs === false) {
        throw new Error('A `>` ref must complement a `<` ref.');
      }
      setTextSelection(editorView, refs['<'], refs['>']);
    }
  }

  const emitTelepointer = (
    from: number,
    to: number = from,
    sessionId = 'guest',
  ) => {
    mockProvider.emit('telepointer', {
      selection: {
        type: 'textSelection',
        anchor: from,
        head: to,
      },
      sessionId,
    });
  };

  type StepType = 'replace' | 'addMark';
  const emitStep = (
    from: number,
    to: number,
    stepType: StepType = 'replace',
  ) => {
    mockProvider.emit('data', {
      json: [{ from, to, stepType }],
    });
  };

  return {
    editorView,
    providerPromise,
    emitTelepointer,
    emitStep,
  };
};

describe.only('Telepointers', () => {
  it('should be able render telepointer on empty document', async () => {
    const {
      editorView,
      providerPromise,
      emitTelepointer,
    } = setupCollabEditor();
    await providerPromise;

    emitTelepointer(1);
    expect(editorView.state).toMatchSnapshot();
  });

  it(`shouldn't remove telepointer`, async () => {
    const {
      editorView,
      providerPromise,
      emitTelepointer,
      emitStep,
    } = setupCollabEditor(doc(p('a{<>}a')));
    await providerPromise;

    emitTelepointer(1);
    emitStep(2, 3);

    expect(editorView.state).toMatchSnapshot();
  });
});
