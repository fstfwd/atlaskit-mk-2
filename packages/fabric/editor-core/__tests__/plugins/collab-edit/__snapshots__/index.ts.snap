// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Telepointers should be able render telepointer on empty document 1`] = `
Object {
  "doc": Object {
    "content": Array [
      Object {
        "type": "paragraph",
      },
    ],
    "type": "doc",
  },
  "selection": Object {
    "anchor": 1,
    "head": 1,
    "type": "text",
  },
  "telepointers": Array [
    Object {
      "from": 1,
      "sessionId": "guest",
      "to": 1,
    },
  ],
}
`;

exports[`Telepointers shouldn't remove telepointer 1`] = `
Object {
  "doc": Object {
    "content": Array [
      Object {
        "content": Array [
          Object {
            "text": "aa",
            "type": "text",
          },
        ],
        "type": "paragraph",
      },
    ],
    "type": "doc",
  },
  "selection": Object {
    "anchor": 2,
    "head": 2,
    "type": "text",
  },
  "telepointers": Array [
    Object {
      "from": 1,
      "sessionId": "guest",
      "to": 1,
    },
  ],
}
`;
